ring/token_range_test.go

```diff
@@ -0,0 +1,306 @@
+package ring
+
+import (
+	"fmt"
+	"math"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestKeyInTokenRanges(t *testing.T) {
+	ranges := TokenRanges{4, 8, 12, 16}
+
+	require.False(t, ranges.IncludesKey(0))
+	require.True(t, ranges.IncludesKey(4))
+	require.True(t, ranges.IncludesKey(6))
+	require.True(t, ranges.IncludesKey(8))
+	require.False(t, ranges.IncludesKey(10))
+	require.False(t, ranges.IncludesKey(20))
+}
+
+func TestGetTokenRangesForInstance(t *testing.T) {
+	numZones := 3
+
+	tests := map[string]struct {
+		zoneTokens map[string][]uint32
+		expected   map[string]TokenRanges
+	}{
+		"single instance in zone": {
+			zoneTokens: map[string][]uint32{
+				"instance-0-0": GenerateTokens(512, nil),
+			},
+			expected: map[string]TokenRanges{
+				"instance-0-0": {0, math.MaxUint32},
+			},
+		},
+		"simple ranges": {
+			zoneTokens: map[string][]uint32{
+				"instance-0-0": {25, 75},
+				"instance-0-1": {10, 50, 100},
+			},
+			expected: map[string]TokenRanges{
+				"instance-0-0": {10, 24, 50, 74},
```

Comment:
This is something that I personally don’t agree with: if token 10 belongs to an instance, I’d expect the corresponding range to include that token

Reply:
I agree with you that it's confusing, and wouldn't mind if we fixed that (in some other PR).

Reply:
I also agree that it's confusing, but it's a function of [this code](https://github.com/grafana/dskit/blob/main/ring/util.go#L129-L140), which causes items that map directly to a token to actually belong to the range owned by the NEXT token.

We could have used half-open ranges in the implementation (`[start, end)`), but that would have made the `IncludesKey` logic more complicated, since we'd need to do an additional check if the `binarySearch` found the key in the ranges slice.

In my opinion, having the token ranges be closed is also easier to reason about, since you don't need to remember the implementation detail of ring instances not actually "owning" any items that map directly to the tokens they have claimed.

Reply:
Note, by "fixing" this, I didn't mean to change how token ranges work. I agree that closed ranges make sense.

What I meant by fixing is modifying logic in `searchToken` method, so that if ingester owns the token, it actually "owns" it :)